### PR TITLE
testing: reduce mocks in AbstractTransportTest to eliminate flakes

### PR DESCRIPTION
**Pull request by ericgribkoff**
The additional mocks that I added in https://github.com/grpc/grpc-java/pull/3145 made at least the `#basicStreams` test flaky: approximately 0.25% of local runs would hit a Mockito `UnfinishedStubbingException`, which I believe was due to the potential for the `doAnswer` blocks to trigger another mocked method. I didn't fully track down the cause, but rather just removed the use of mocks for `ClientStreamListener` and `ServerStreamListener` altogether, and no longer observe failures across many thousands of runs.

This also adds an `#awaitOutboundHeaders` method to `TestClientStreamTracer` (https://github.com/grpc/grpc-java/pull/3305), as without this a small percentage of runs of `#basicStreams` would also fail due to a race between the server receiving the headers and the Netty future on the client invoking the stream tracer context to report headers sent.

### CODE REVIEW

File: testing/src/main/java/io/grpc/internal/testing/TestClientStreamTracer.java [hunk 1]
```diff
@@ -26,6 +27,7 @@
  */
 public class TestClientStreamTracer extends ClientStreamTracer implements TestStreamTracer {
   private final TestBaseStreamTracer delegate = new TestBaseStreamTracer();
+  private final CountDownLatch outboundHeadersLatch = new CountDownLatch(1);
```
Comment (carl-mastrangelo) [inline testing/src/main/java/io/grpc/internal/testing/TestClientStreamTracer.java]: consider making this protected.  The rest of this class is designed to be overridable.  

For example the `outboundHeaders()` method below could be overridden,  but have no way to count down this latch.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 2]
```diff
@@ -1622,14 +1486,114 @@ public StreamCreation takeStreamOrFail(long timeout, TimeUnit unit)
...
+      }
+    }
+
+    @Override
+    public void onReady() {
+      if (status.isDone()) {
+        fail("onReady invoked after closed");
+      }
+      onReadyLatch.countDown();
+      ;
```
Comment (carl-mastrangelo) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: extra semi

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 3]
```diff
@@ -1622,14 +1486,114 @@ public StreamCreation takeStreamOrFail(long timeout, TimeUnit unit)
...
```
Comment (carl-mastrangelo) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: same here

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 2 above)
Comment (ericgribkoff) [reply testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Fixed

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 3 above)
Comment (ericgribkoff) [reply testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Fixed

File: testing/src/main/java/io/grpc/internal/testing/TestClientStreamTracer.java (hunk 1 above)
Comment (ericgribkoff) [reply testing/src/main/java/io/grpc/internal/testing/TestClientStreamTracer.java]: Done

**Pull request merged**